src/librbd/ExclusiveLock.cc

```diff
@@ -132,610 +93,242 @@ void ExclusiveLock<I>::init(uint64_t features, Context *on_init) {
 
 template <typename I>
 void ExclusiveLock<I>::shut_down(Context *on_shut_down) {
-  ldout(m_image_ctx.cct, 10) << this << " " << __func__ << dendl;
+  ldout(m_image_ctx.cct, 10) << dendl;
 
-  {
-    Mutex::Locker locker(m_lock);
-    assert(!is_shutdown());
-    execute_action(ACTION_SHUT_DOWN, on_shut_down);
-  }
+  ML<I>::shut_down(on_shut_down);
 
   // if stalled in request state machine -- abort
   handle_peer_notification();
 }
 
 template <typename I>
 void ExclusiveLock<I>::try_lock(Context *on_tried_lock) {
```

Comment:
Minor: it would be great, while you are in here, to just remove the ```try_lock``` method and '''request_lock``` method and modify librbd to use ```try_acquire_lock``` and ```acquire_lock``` for simplicity / clarity.